File: backend/services/stigmer-server/pkg/domain/mcpserver/oauth/managed_env.go
package oauth

import (
	"context"
	"fmt"

	"github.com/rs/zerolog/log"
	environmentv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/environment/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/downstream/environment"
)

const managedEnvLabel = "stigmer.ai/managed"

// ManagedEnvironmentService creates and manages system-controlled environments
// for storing OAuth tokens. Each managed environment is scoped to a single
// (user, resource, org) tuple and identified by the stigmer.ai/managed=true label.
//
// All operations go through the downstream environment gRPC client, which
// provides encryption, validation, and audit automatically via the environment
// pipeline. This is a deliberate cross-domain access, analogous to the
// OAuthAppRepo access in OAuthTokenRefreshService.
type ManagedEnvironmentService struct {
	environmentClient *environment.Client
}

// NewManagedEnvironmentService creates a new service backed by the given
// downstream environment client.
func NewManagedEnvironmentService(envClient *environment.Client) *ManagedEnvironmentService {
	return &ManagedEnvironmentService{environmentClient: envClient}
}

// CreateManagedEnvironment creates a new system-managed environment with the
// stigmer.ai/managed=true label. Returns the environment's resource ID.
//
// The environment goes through the standard create pipeline, which handles
// ID generation, slug resolution, timestamps, and search indexing.
func (s *ManagedEnvironmentService) CreateManagedEnvironment(
	ctx context.Context,
	name string,
	org string,
) (string, error) {
	created, err := s.environmentClient.Create(ctx, &environmentv1.Environment{
		ApiVersion: "agentic.stigmer.ai/v1",
		Kind:       "Environment",
		Metadata: &apiresource.ApiResourceMetadata{
			Name: name,
			Org:  org,
			Labels: map[string]string{
				managedEnvLabel: "true",
			},
		},
	})
	if err != nil {
		return "", fmt.Errorf("failed to create managed environment: %w", err)
	}

	envID := created.GetMetadata().GetId()

	log.Info().
		Str("environment_id", envID).
		Str("name", name).
		Str("org", org).
		Msg("Created managed environment for OAuth token storage")

	return envID, nil
}

// DeleteManagedEnvironment deletes a managed environment and all its secrets.
// The environment pipeline handles cleanup of encryption keys, labels, and
// search index entries.
//
// Non-fatal callers should catch errors — the environment may already be
// deleted (e.g., concurrent disconnect calls or partial failure retries).
func (s *ManagedEnvironmentService) DeleteManagedEnvironment(
	ctx context.Context,
	environmentID string,
) error {
	_, err := s.environmentClient.Delete(ctx, &apiresource.ApiResourceDeleteInput{
		ResourceId: environmentID,
	})
	if err != nil {
		return fmt.Errorf("failed to delete managed environment %s: %w", environmentID, err)
	}

	log.Info().
		Str("environment_id", environmentID).
		Msg("Deleted managed environment")

	return nil
}

// UpdateSecrets writes secret variables into a managed environment.
// Values must be plaintext (never enc:v<N>:-prefixed — the environment
// pipeline rejects ciphertext-shaped input, oss#395) and are stored as
// passed: the OSS environment pipeline does not encrypt at write time
// (at-rest encryption for environment values is tracked separately).
func (s *ManagedEnvironmentService) UpdateSecrets(
	ctx context.Context,
	environmentID string,
	variables map[string]*environmentv1.EnvironmentValue,
) error {
	_, err := s.environmentClient.UpdateVariables(ctx, &environmentv1.UpdateEnvironmentVariablesRequest{
		EnvironmentId: environmentID,
		Variables:     variables,
	})
	if err != nil {
		return fmt.Errorf("failed to update secrets in managed environment %s: %w", environmentID, err)
	}
	return nil
}

// ReadSecretValue retrieves a single decrypted secret value from a managed
// environment. Returns the plaintext value.
func (s *ManagedEnvironmentService) ReadSecretValue(
	ctx context.Context,
	environmentID string,
	key string,
) (string, error) {
	val, err := s.environmentClient.GetSecretValue(ctx, &environmentv1.EnvironmentSecretValueInput{
		EnvironmentId: environmentID,
		Key:           key,
	})
	if err != nil {
		return "", fmt.Errorf("failed to read secret %q from managed environment %s: %w", key, environmentID, err)
	}
	return val.GetValue(), nil
}

File: backend/services/stigmer-server/src/domain/mcpserver/oauth/managed-env.ts
/**
 * Managed-environment access — ports pkg/domain/mcpserver/oauth/
 * managed_env.go whole: reading and rewriting OAuth token secrets (#17's
 * EC-builder slice) plus creating and deleting the managed environments
 * themselves (#19's connect/OAuth slice). Every operation rides the
 * environment domain's in-process client, so encryption, validation, and
 * audit come from the environment pipeline automatically (DD-002: full
 * interceptor traversal on every internal call).
 */
import { create } from "@bufbuild/protobuf";

import type { EnvironmentValue } from "@stigmer/protos/ai/stigmer/agentic/environment/v1/spec_pb";
import type { UpdateEnvironmentVariablesRequest } from "@stigmer/protos/ai/stigmer/agentic/environment/v1/io_pb";
import {
  EnvironmentSecretValueInputSchema,
  UpdateEnvironmentVariablesRequestSchema,
} from "@stigmer/protos/ai/stigmer/agentic/environment/v1/io_pb";
import type { Environment } from "@stigmer/protos/ai/stigmer/agentic/environment/v1/api_pb";
import { EnvironmentSchema } from "@stigmer/protos/ai/stigmer/agentic/environment/v1/api_pb";
import { ApiResourceDeleteInputSchema } from "@stigmer/protos/ai/stigmer/commons/apiresource/io_pb";
import type { MessageInitShape } from "@bufbuild/protobuf";

import type { Logger } from "../../../boot/logger.js";
import type { CallerIdentity } from "../../../extensions/identity.js";

/** The label marking system-managed OAuth-token environments. */
export const MANAGED_ENV_LABEL = "stigmer.ai/managed";

/**
 * The narrow in-process environment surface this service consumes —
 * satisfied by the composition root's in-process clients (DD-002: full
 * interceptor traversal on every internal call).
 */
export interface ManagedEnvironmentClient {
  getSecretValue(
    input: MessageInitShape<typeof EnvironmentSecretValueInputSchema>,
  ): Promise<EnvironmentValue>;
  updateVariables(
    request: MessageInitShape<typeof UpdateEnvironmentVariablesRequestSchema>,
  ): Promise<Environment>;
  /**
   * `caller` propagates the ORIGINAL caller through the in-process hop
   * (ruling R5; Java createAsCaller — parity entry 20260830.05): the
   * created environment's owner attribution lands on the connecting
   * user, so it stays visible and manageable under an enforcing
   * Authorizer. Absent = the minted internal class (the pre-R5 shape,
   * kept for hops with no request caller).
   */
  create(
    environment: MessageInitShape<typeof EnvironmentSchema>,
    caller?: CallerIdentity,
  ): Promise<Environment>;
  delete(
    input: MessageInitShape<typeof ApiResourceDeleteInputSchema>,
  ): Promise<Environment>;
}

export class ManagedEnvironmentService {
  constructor(
    private readonly client: ManagedEnvironmentClient,
    private readonly logger: Logger,
  ) {}

  /**
   * Creates a new system-managed environment with the
   * stigmer.ai/managed=true label and returns its resource id (Go
   * CreateManagedEnvironment). The environment goes through the standard
   * create pipeline, which handles id generation, slug resolution,
   * timestamps, and search indexing. `caller` propagates the connecting
   * user through the hop (ruling R5, parity entry 20260830.05) so a
   * composed tuple-lifecycle driver attributes ownership to them — the
   * Java createAsCaller posture ("org link + owner = caller"); the
   * reserved-label guard keys its trust arm on the in-process ORIGIN,
   * so the managed label still passes with a propagated caller.
   */
  async createManagedEnvironment(
    name: string,
    org: string,
    caller?: CallerIdentity,
  ): Promise<string> {
    let created: Environment;
    try {
      created = await this.client.create(
        create(EnvironmentSchema, {
          apiVersion: "agentic.stigmer.ai/v1",
          kind: "Environment",
          metadata: {
            name,
            org,
            labels: { [MANAGED_ENV_LABEL]: "true" },
          },
        }),
        caller,
      );
    } catch (error) {
      throw new Error(
        `failed to create managed environment: ${error instanceof Error ? error.message : String(error)}`,
      );
    }

    const envId = created.metadata?.id ?? "";
    this.logger.info("Created managed environment for OAuth token storage", {
      environment_id: envId,
      name,
      org,
    });
    return envId;
  }

  /**
   * Deletes a managed environment and all its secrets (Go
   * DeleteManagedEnvironment). Non-fatal callers should catch errors —
   * the environment may already be deleted (concurrent disconnects or
   * partial-failure retries).
   */
  async deleteManagedEnvironment(environmentId: string): Promise<void> {
    try {
      await this.client.delete(
        create(ApiResourceDeleteInputSchema, { resourceId: environmentId }),
      );
    } catch (error) {
      throw new Error(
        `failed to delete managed environment ${environmentId}: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
    this.logger.info("Deleted managed environment", {
      environment_id: environmentId,
    });
  }

  /**
   * Retrieves a single decrypted secret value from a managed environment
   * (Go ReadSecretValue).
   */
  async readSecretValue(environmentId: string, key: string): Promise<string> {
    let value: EnvironmentValue;
    try {
      value = await this.client.getSecretValue(
        create(EnvironmentSecretValueInputSchema, {
          environmentId,
          key,
        }),
      );
    } catch (error) {
      throw new Error(
        `failed to read secret ${JSON.stringify(key)} from managed environment ${environmentId}: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
    return value.value;
  }

  /**
   * Writes secret variables into a managed environment (Go
   * UpdateSecrets). Values must be plaintext (never enc:vN:-prefixed —
   * the environment pipeline rejects ciphertext-shaped input, oss#395);
   * the pipeline encrypts is_secret values at rest (oss#405).
   */
  async updateSecrets(
    environmentId: string,
    variables: { [key: string]: EnvironmentValue },
  ): Promise<void> {
    try {
      await this.client.updateVariables(
        create(UpdateEnvironmentVariablesRequestSchema, {
          environmentId,
          variables,
        }),
      );
    } catch (error) {
      throw new Error(
        `failed to update secrets in managed environment ${environmentId}: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }
}
